keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -32,218 +31,103 @@ status: provisional
-
-#### Trace Utility Package
-
-This package will be able to create spans from the span context embedded in the `trace.kubernetes.io/context` object annotation, in addition to embedding context from spans back into the annotation. This package will facilitate propagating traces through kubernetes objects.  The exported functions include:
-
-```golang
-// InitializeExporter initializes the trace exporting service with the provided service name.
-// Components should use this initializer to ensure common behavior.
-func InitializeExporter(service string)
-
-// StartSpanFromObject constructs a new Span using the context attached to the object as the parent SpanContext.  It mirrors trace.StartSpan, but for kubernetes objects.
-func StartSpanFromObject(ctx context.Context, obj meta.Object, spanName string) (context.Context, *trace.Span, error)
-
-// EncodeContextIntoObject encodes the SpanContext contained in the context into the object
-func EncodeContextIntoObject(ctx context.Context, obj meta.Object)
-
-// RemoveSpanContextFromObject removes the SpanContext attached to an object, if one exists
-func RemoveSpanContextFromObject(obj meta.Object) 
-```
-
-#### Tracing Pod Lifecycle
+The API Server will use the [OpenTelemetry exporter format](https://github.com/open-telemetry/opentelemetry-proto), which exports traces to a local port.  This format is compatible with the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector), which allows importing and configuring exporters for trace storage backends to be done out-of-tree in addition to other useful features.  The exporter stores spans in memory, and uses the [batching processor](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#batching-processor) to batch requests and send them asynchronously.
 
-As we move forward with this KEP, we will use the aforementioned trace utility package to trace pod-related operations across the scheduler and kubelet. In code, this corresponds to creating a span (i.e. `ctx, span := trace.StartSpan(ctx, "Component.SampleSpan")`) at the beginning of an operation, and ending the span afterwards (`span.End()`). All calls to tracing functions will be gated with the `ObjectLifecycleTracing` alpha feature-gate, and will be disabled by default.
+Add configuration to the API Server required to configure the opentelemetry exporter, including the address and egress proxy to send spans to. The [egress proxy](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190226-network-proxy.md) can be added to the opentelemetry exporter by adding a ContextDialer grpc DialOption similar to the one used by the apiserver's etcd client.  This will add a new "OpenTelemetry" [EgressType](https://github.com/kubernetes/kubernetes/blob/4b9b9ab75376b7b53876ab6b2be42d0940c7eb26/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go#L53) to the API Server's configuration.
```

Comment:
Can you explain why it requires a new EgressType? If the egress proxy is in the same network as the apiserver, can you just use the Master egressType?

Reply:
That would work, although it wouldn't be quite as flexible.

The OpenTelemetry collector is generally run as both a daemonset and a deployment + service in a cluster.  The daemonset collects from each node, and sends to the deployment + service for aggregation.  So to add the collector to the master, we have two options:

1. Run the OTel daemonset on the master node as well.
2. Configure master components to send directly to the service + deployment.

(1) is generally preferred, but means the daemonset has to be able to send to an in-cluster service.  I was able to get this working by running kube-proxy on the master on a kubetest cluster on gce.  But I don't think that works when the master and cluster are on different networks.

I suppose it would be possible to write an [OTel exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/master/exporter) that can send over the proxy instead of having a different egress type.

One advantage of using the Master egress type, is that if any other master components want to send telemetry using OpenTelemetry, having the agent already working on the master would make those work out of the box.  If the API server was configured to send to the cluster service, all other master components would need to be able to send to the proxy to work.

Let me know your thoughts, and i'll update the KEP, and add an Alternative Considered depending on which way we go.

Reply:
> 1. Run the OTel daemonset on the master node as well.
> 2. Configure master components to send directly to the service + deployment.

IIUC, If we pick 1, then the API server sends traffic to a daemonset running on the master node, then the traffic will be of the "Master" EgressType. If we pick 2, then the API server sends traffic directly to the "deployment+service" running in the cluster, so the traffic will be of the "Cluster" EgressType.

I still don't get why we need an additional "OpenTelemetry" EgressType and how that's more flexible.


> I suppose it would be possible to write an OTel exporter that can send over the proxy instead of having a different egress type.

what do you mean by "proxy" here? Are you referring to the [apiserver-networking-proxy](https://github.com/kubernetes-sigs/apiserver-network-proxy)?

Also I'm confused by what an OTel exporter is. Is it a standalone process? Or is it going to be part of the API server? I.e., will the traffic be like:
a. apiserver -> OTel exporter -> OTel collector
or
b. apiserver (which calls the OTel exporter library) -> OTel collector?

Reply:
> (1) is generally preferred, but means the daemonset has to be able to send to an in-cluster service. I was able to get this working by running kube-proxy on the master on a kubetest cluster on gce. But I don't think that works when the master and cluster are on different networks.

You can let the daemonset sends to the traffic to the [apiserver-network-proxy](https://github.com/kubernetes-sigs/apiserver-network-proxy), which can tunnel the traffic to the cluster.

Reply:
> I still don't get why we need an additional "OpenTelemetry" EgressType and how that's more flexible.

We wouldn't have to pick either Cluster or Master for everyone; we could do either.

> what do you mean by "proxy" here? Are you referring to the apiserver-networking-proxy?

Yes, i'm referring to the apiserver-networking-proxy.

> Also I'm confused by what an OTel exporter is. Is it a standalone process? Or is it going to be part of the API server? I.e., will the traffic be like: [a. or b.]

It will be:
c. apiserver -> Otel collector (daemonset) w/ custom exporter that supports proxy -> Otel collector (deployment + service)

An exporter is a compiled-in plugin that can be added to the collector.

> You can let the daemonset sends to the traffic to the apiserver-network-proxy, which can tunnel the traffic to the cluster.

Yeah, thats what i'm suggesting as well with "I suppose it would be possible to write an OTel exporter that can send over the proxy instead of having a different egress type."  My understanding is that having the daemonset send over the apiserver-network-proxy requires modifying the daemonset's client.  I'm just suggesting we do that in plugin code.

Reply:
> An exporter is a compiled-in plugin that can be added to the collector.

That's the piece that I was missing. Thanks for the detailed explanations.

So we have two choices so far:
1. add an "OpenTelemetry" egressType, and by configuring this egressType, users can control if the apiserver sends the trace data to the collector running on the master node or the collector running in the cluster.
2. always let the apiserver send tracing data to the collector that's running on the master node. And if the collector needs to send the data to the collectors running in the cluster, it can call the exporter, which is modified to call the apiserver-network-proxy to tunnel the traffic.

I think we have a third option that's more conventional. How do you configure the apiserver client that's talking to the collector? We can do that via a configuration object, say `OpenTelemetryClientConfiguration", and make it similar to the [webhook client configuration](https://github.com/kubernetes/kubernetes/blob/b9d2df810c89bde69cd790905ec937caf1c36d19/staging/src/k8s.io/api/admissionregistration/v1/types.go#L488).

```
type OpenTelemetryClientConfiguration struct {
	// +optional
	// URL of the collector that's running on the master.
       // if URL is specified, APIServer uses the egressType Master when sending tracing data to the collector.
	URL *string `json:"url,omitempty" protobuf:"bytes,3,opt,name=url"`

	// +optional
	// Service that's the frontend of the collector deployment running in the cluster.
	// If Service is specified, APIServer uses the egressType Cluster when sending tracing data to the collector.
	Service *ServiceReference `json:"service,omitempty" protobuf:"bytes,1,opt,name=service"`
}
```

Reply:
That makes a lot of sense.  It looks like those configurations are actually stored as kubernetes objects.  I would think we wouldn't want to expose this configuration outside of the master node, right?  Would this be part of a file-based configuration instead?

My other thought is that if we decide the third option (add a new config with URL or Service) is the best, we may want to delay this until after the initial implementation to gather feedback and use-cases for a service-based setup.  I.E. We always use "Master" egress type to start, and can add configuration for a "Cluster" based egress in the future if we have compelling use-cases.

Reply:
> I would think we wouldn't want to expose this configuration outside of the master node, right? Would this be part of a file-based configuration instead?

Yeah, we can start with file-based configurations. We can always promote it to a public API if we want to.

> use-cases for a service-based setup. I.E. We always use "Master" egress type to start, 

For service-based setup, it should be "Cluster" egress type.

I agree that we can add the `OpenTelemetryClientConfiguration` configuration after we get initial feedback.

Reply:
I updated the KEP based on this discussion.